src/liana/method/sp/_compute_global_specificity.py

```diff
@@ -0,0 +1,208 @@
+
+    # Build sparse one-hot and normalize columns
+    t_sparse = csr_matrix(
+        (np.ones(n_cells, dtype=np.float64), (row_idx, col_idx)),
+        shape=(n_cells, n_types)
+    )
+    col_sums = np.asarray(t_sparse.sum(axis=0)).ravel()
+    col_sums[col_sums == 0] = 1.0
+    t_sparse = t_sparse @ diags(1.0 / col_sums)
+
+    # Ensure X is sparse CSR
+    X = X if issparse(X) else csr_matrix(X)
+
+    # Aggregation
+    result = t_sparse.T @ X
+    values = (result.toarray() if issparse(result) else np.asarray(result)).ravel()
+
+    # Build keys with vectorized string ops
+    interaction_names_tiled = np.tile(interaction_names.astype(str), n_types)
+    celltype_names_repeated = np.repeat(
+        np.asarray(celltype_names, dtype=str),
+        interaction_names.shape[0]
+    )
+    keys = np.char.add(np.char.add(interaction_names_tiled, lr_sep), celltype_names_repeated)
+
+    return dict(zip(keys.tolist(), values.tolist(), strict=True))
+
+@d.dedent
+def compute_global_specificity(
+    adata: AnnData,
+    groupby: str,
+    lr_sep: str = V.lr_sep,
+    complex_sep: str = V.complex_sep,
+    n_perms: int = V.n_perms,
+    seed: int = V.seed,
+    n_jobs: int = -1,
+    verbose: bool = V.verbose,
+) -> None:
+    """
+    Computes global specificity and calculates permutation test p-values.
+
+    Args:
+        %(adata)s
+        %(groupby)s
+        lr_sep (str, optional): Separator for names. Defaults to `V.lr_sep` ('^').
+        complex_sep (str, optional): Separator for splitting complex names. Defaults to "_".
+        %(n_perms)s
+        %(seed)s
+        %(verbose)s
+
+    Returns
+    -------
+        None: The result with 'lr_mean' and 'pval' is stored in `adata.uns["global_interactions"]`.
+    """
+    if groupby not in adata.obs.columns:
+        raise KeyError(
+            f"`groupby`='{groupby}' not found in adata.obs. "
+            "Use the same grouping column used to build adata."
+        )
+
+    rng_main = np.random.default_rng(seed)
+    original_groupby_labels = adata.obs[groupby].copy()
+    
+    # --- Part A: Observed score (sparse) ---
+
+    # Fixed column order for cell types
+    celltypes = pd.get_dummies(adata.obs[groupby])
+    celltype_names = list(celltypes.columns)
+
+    # Map original labels to fixed indices
+    cats_obs = pd.Categorical(adata.obs[groupby].values, categories=celltype_names, ordered=False)
+    col_idx_obs = cats_obs.codes
+
+    n_cells = col_idx_obs.size
+    n_types = len(celltype_names)
+    row_idx = np.arange(n_cells)
+
+    # Sparse one-hot and normalize columns
+    t_sparse = csr_matrix(
+        (np.ones(n_cells, dtype=np.float64), (row_idx, col_idx_obs)),
+        shape=(n_cells, n_types)
+    )
+    col_sums = np.asarray(t_sparse.sum(axis=0)).ravel()
+    col_sums[col_sums == 0] = 1.0
+    t_sparse = t_sparse @ diags(1.0 / col_sums)
+
+    # Ensure X is sparse CSR
+    X_raw = adata.X
+    X = X_raw if issparse(X_raw) else csr_matrix(X_raw)
```

Comment:
This should instead use: https://github.com/saezlab/liana-py/blob/ea90b7561c549dbd149c0297e28d7ea7d903b55f/src/liana/method/_pipe_utils/_pre.py#L222